src/Codegen/Codegen.hs

```diff
@@ -586,72 +586,92 @@ genPrim I.Return [e] _ = do
+genLiteral (I.LitBool     True ) _ = return (marshal [cexp|true|], [])
+genLiteral (I.LitBool     False) _ = return (marshal [cexp|false|], [])
 genLiteral I.LitEvent            _ = return ([cexp|1|], [])
 
 -- | Generate C expression for SSM primitive operation.
 genPrimOp :: PrimOp -> [Expr] -> Type -> GenFn (C.Exp, [C.BlockItem])
 genPrimOp I.PrimAdd [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
-  return ([cexp|$exp:lhsVal + $exp:rhsVal|], stms)
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
+  -- all integers are 31 bits + 1 tag bit, so zero tag bit on one argument,
+  -- add together, and the result will be sum with a tag bit of 1.
+  return ([cexp|((((typename uint32_t) $exp:rhsVal) & 0xFFFFFFFE) + ((typename uint32_t) $exp:lhsVal))|], stms)
 genPrimOp I.PrimSub [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
-  return ([cexp|$exp:lhsVal - $exp:rhsVal|], stms)
+  -- all integers are 31 bits + 1 tag bit, so zero tag bit on subtrahend,
+  -- subtract, and the result will be difference with a tag bit of 1.                   
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
+  return ([cexp|(((typename uint32_t) $exp:lhsVal) - (((typename uint32_t) $exp:rhsVal) & 0xFFFFFFFE))|], stms)
 genPrimOp I.PrimMul [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
-  return ([cexp|$exp:lhsVal * $exp:rhsVal|], stms)
+  ((lhsVal, rhsVal), stms) <-
+    first (bimap unmarshal unmarshal) <$> genBinop lhs rhs
+  return (marshal [cexp|$exp:lhsVal * $exp:rhsVal|], stms)
 genPrimOp I.PrimDiv [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
-  return ([cexp|$exp:lhsVal / $exp:rhsVal|], stms)
+  ((lhsVal, rhsVal), stms) <-
+    first (bimap unmarshal unmarshal) <$> genBinop lhs rhs
+  return (marshal [cexp|$exp:lhsVal / $exp:rhsVal|], stms)
 genPrimOp I.PrimMod [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
-  return ([cexp|$exp:lhsVal % $exp:rhsVal|], stms)
+  ((lhsVal, rhsVal), stms) <-
+    first (bimap unmarshal unmarshal) <$> genBinop lhs rhs
+  return (marshal [cexp|$exp:lhsVal % $exp:rhsVal|], stms)
 genPrimOp I.PrimNeg [opr] _ = do
-  (val, stms) <- genExpr opr
-  return ([cexp|- $exp:val|], stms)
+  (val, stms) <- first unmarshal <$> genExpr opr
+  return (marshal [cexp|- $exp:val|], stms)
 genPrimOp I.PrimBitAnd [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal & $exp:rhsVal|], stms)
 genPrimOp I.PrimBitOr [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal | $exp:rhsVal|], stms)
 genPrimOp I.PrimBitNot [opr] _ = do
   (val, stms) <- genExpr opr
-  return ([cexp|~ $exp:val|], stms)
+  -- all integers are 31 bits + 1 tag bit, so val XOR (~1)
+  -- flips the 31 bits and keeps the tag bit 1. 
+  return ([cexp|($exp:val ^ 0xFFFFFFFE)|], stms)
```

Comment:
nit: why don't we change `0xFFFFFFFE` to `~1`?

Reply:
Since ~1 is a constant, I thought I could save an assembly instruction by writing 0xFFFFFFFE. I can change it to ~1 if it's clearer though.

Reply:
An optimizing compiler will always fold constants like this; the problem with using 0xFFFF.. is that it only works with 32-bit integers and will break on a 64-bit platform, which we should keep in mind.

Reply:
I used 0xFFFFFFFE in other parts of the generated C code., like in genPrimOp I.PrimAdd. So I should change all instances of 0xFFFFFFFE  to ~1 to ensure compatibility, right?